File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboSeries.java
package com.hiddenramblings.tagmo.amiibo;

import androidx.annotation.NonNull;

public class AmiiboSeries implements Comparable<AmiiboSeries> {
    public static final long MASK = 0x000000000000FF00L;
    public static final int BITSHIFT = 4 * 2;

    public final AmiiboManager manager;
    public final long id;
    public final String name;

    public AmiiboSeries(AmiiboManager manager, long id, String name) {
        this.manager = manager;
        this.id = id;
        this.name = name;
    }

    public AmiiboSeries(AmiiboManager manager, String id, String name) {
        this(manager, hexToId(id), name);
    }

    public static long hexToId(String value) {
        return (Long.decode(value) << BITSHIFT) & MASK;
    }

    @Override
    public int compareTo(@NonNull AmiiboSeries amiiboSeries) {
        return this.name.compareTo(amiiboSeries.name);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboSeries.kt
package com.hiddenramblings.tagmo.amiibo

class AmiiboSeries(val manager: AmiiboManager, val id: Long, val name: String) :
    Comparable<AmiiboSeries> {
    constructor(manager: AmiiboManager, id: String, name: String) : this(
        manager, hexToId(id), name
    )

    override fun compareTo(other: AmiiboSeries): Int {
        return name.compareTo(other.name)
    }

    companion object {
        const val MASK = 0x000000000000FF00L
        const val BITSHIFT = 4 * 2
        fun hexToId(value: String): Long {
            return java.lang.Long.decode(value) shl BITSHIFT and MASK
        }
    }
}